xarray/core/common.py

```diff
@@ -115,7 +120,15 @@ def __iter__(self):
     def T(self):
         return self.transpose()
 
-    def get_axis_num(self, dim):
+    @overload
+    def get_axis_num(self, dim: str) -> int:
+        pass
+
+    @overload  # noqa:F811
+    def get_axis_num(self, dim: Iterable[str]) -> Tuple[int]:
```

Comment:
Dimensions are actually allowed to be non-strings, so it might be better to keep this overload more generic, given that that reflects how the code actually works:
```suggestion
    def get_axis_num(self, dim: Iterable[Hashable]) -> Tuple[int, ...]:
```

(Also `Tuple[int]`  should be `Tuple[int, ...]`)

Reply:
If they were intended to be allowed to be non-strings, sure they aren't tested for it!
```
>>> xarray.DataArray([1,2], dims=[123])
TypeError: dimension 123 is not a string
```
Also I've never seen it mentioned in the documentation...

Reply:
fixed ``Tuple[int, ...]``

Reply:
Hmm. It seems that you can only construct these sorts of names/dimensions with the Dataset API:
```
In [8]: xarray.Dataset({123: ((456,), [1, 2, 3])})
Out[8]:
<xarray.Dataset>
Dimensions:  (456: 3)
Dimensions without coordinates: 456
Data variables:
    123      (456) int64 1 2 3
```